### PR TITLE
Allowing references in R Markdown to PDF vignettes

### PR DESCRIPTION
**Pull request by grimbough**
This primarily addresses #94 where citations can't be used inside an R Markdown vignette using `BiocStyle::pdf_document()`

I've added a new pandoc template that includes the missing latex environments `CSLReferences` and `cslreferences` required by **citeproc**.  I've passed this on similar features found in the **tufte** package.  Interestingly **rmarkdown** now seems to use the templates shipped with pandoc, and maybe that would also be a way to deal with this, but the current approach seems to work.

I've also added a GitHub workflow to build the R Markdown vignette with various versions of Pandoc and TinyTex.  I've also added a citation to the Rmd vignette so that the referencing infrastructure can be tested.  This workflow would fail before my patch here is applied.

---

I've also made three small changes to `Bioconductor.sty` to suppress some warnings that are always produced, but which I don't think are informative to an end user as they can't do anything to change them.  These were:

```
Warning messages:
LaTeX Warning: You have requested package `/BiocStyle/resources/tex/Bioconductor',
               but the package provides `Bioconductor'.
Package Fancyhdr Warning: \fancyhead's `E' option without twoside option is use less on input line 173.
Package geometry Warning: Over-specification in `h'-direction. `width' (384.1122pt) is ignored.
```

I can restore the if they may be useful, but I've always been slightly annoyed to see them.  
~I don't see an easy way to fix the first error, so I just suppress it with the silence package.~  *I've reverted this, it was breaking the Sweave vignette.*
The second might influence the layout of a two-sided PDF, but I'm not sure it's actually possible to produce one of those with BiocStyle. 
 The third states it is ignored, so I guess it doesn't do anything and I've removed the line that leads to it.

### CODE REVIEW

**Comment by aoles** [comment]
Thanks a lot @grimbough for addressing the issue with citations in PDF output from RMD vignettes, and all the rest of small fixes included in this PR. I'm happy to review it whenever you're ready, just let me know. Cheers!

**Comment by grimbough** [comment]
Thanks @aoles I'd happy recieve a review now.

The second set of commits address an issue I've noticed recently where newer versions of pandoc add some additional classes to the footnote `div` and the existing code doesn't find them and move them to the margins.  I've added a unit tests and a github workflow that uses a variety of Pandoc versions to try and tests for this.  You can see and R CMD check that fails with Pandoc 2.17 at https://github.com/grimbough/BiocStyle/actions/runs/2436503421 and then after my patch at https://github.com/grimbough/BiocStyle/actions/runs/2436503421

**Review by aoles — COMMENTED** [review]
Hi Mike!

Thanks again for the improvements to the handling of citations in BiocStyle. After a quick look I have the following initial ideas and suggestions:

1. I was a bit surprised to see that by default pandoc-generated citations are not hyperlinked. As all of the other references such as to figures, tables and equations, as well as footnotes are linked, I think a reasonable thing would be to enable citations to render as links as well. For the html output it might be enough to have the YAML parameter `link-citations: yes` set by default. As for the pdf output, I'm not sure what the best approach would be, but I believe this should be feasible too. Is this something you would be willing to look into?
2. Also, the way how the bibliography in the "Authoring R Markdown vignettes" vignette shows up appended directly after the output of `sessionInfo()` looks somehow weird to me. Do you think you could improve on this? What I could imagine is a similar structure as in the "Bioconductor LATEX Style 2.0" vignette: have "References" as an unnumbered section, and move "Session info" to an Appendix, cf. [here](https://bookdown.org/yihui/rmarkdown-cookbook/bibliography.html#include-appendix-after-bibliography). 

Cheers,
Andrzej

**Comment by grimbough** [comment]
Thanks for the review @aoles, sorry I haven't replied until now.   I'm currently on parerntal leave for a few months and had too many things to tidy up before I stopped work.  I'll happily take a look at making the suggested changes when I get some time, but it might be a while.

**Comment by grimbough** [comment]
> 2\. Also, the way how the bibliography in the "Authoring R Markdown vignettes" vignette shows up appended directly after the output of `sessionInfo()` looks somehow weird to me. Do you think you could improve on this? What I could imagine is a similar structure as in the "Bioconductor LATEX Style 2.0" vignette: have "References" as an unnumbered section, and move "Session info" to an Appendix, cf. [here](https://bookdown.org/yihui/rmarkdown-cookbook/bibliography.html#include-appendix-after-bibliography).

Hi @aoles For the point above, do we achieve the formatting programmatically, or is just that the default template for the BiocStyle sweave format has a better example layout and so achieve a nicer end product, but someone could still make an ugly document if they tried?

**Review by aoles — CHANGES_REQUESTED** [review]
Hi Mike,

Long time no see! I hope you enjoyed your parental leave. Let me apologize for not getting back to you earlier, the end of last year was quite a busy time for me. 

Thanks a lot for working on this and for addressing my comments. This looks really great now, keep up the good work! I still have some minor improvement suggestions, please find them in the inline comments.

Cheers!
Andrzej

File: NEWS [hunk 1]
```diff
@@ -1,3 +1,15 @@
+CHANGES IN VERSION 2.26.0
```
Comment (aoles) [inline NEWS:1]: Requires version bump.

**Comment by aoles** [comment]
Hi Mike @grimbough,

I hope you are doing fine. I was wondering whether it would be feasible to wrap-up the review soonish in order to merge before the upcoming BioC release? Please let me know if there is anything I can help you with.

Looking forward to hearing from you.

Cheers!
Andrzej

File: inst/rmarkdown/templates/html_document/skeleton/skeleton.Rmd [hunk 2]
```diff
@@ -15,6 +15,7 @@ vignette: |
   %\VignetteIndexEntry{Vignette Title}
   %\VignetteEngine{knitr::rmarkdown}
   %\VignetteEncoding{UTF-8}
+link-citations: yes
```
[the remark this answers is not in the document]
Comment (grimbough) [reply inst/rmarkdown/templates/html_document/skeleton/skeleton.Rmd]: I'm not sure how to do this.  I can't figure out where the `link-citations` setting the YAML header gets processed by bookdown/rmarkdown/knitr.  All the references I can find to that keyword in those packages just have it set in the YAML header of example files, but not processed by code.  I also don't see it passed as an argument to pandoc when I'm compiling the document.  Is it possible it's something special that Pandoc requires in the document header at conversion time?

[the remark this answers is not in the document]
Comment (aoles) [reply inst/rmarkdown/templates/html_document/skeleton/skeleton.Rmd]: Thanks for getting back to me! To be honest, I haven't looked into it myself, but rather assumed this should be doable in a similar fashion as with other document options.

File: vignettes/AuthoringRmdVignettes.Rmd [hunk 3]
```diff
@@ -372,8 +373,20 @@ Footnotes are displayed as side notes on the right margin^[this is a side note
 entered as a footnote], which has the advantage that they appear close to the
 place where they are defined.
 
+# Bibliography
+
+If you wish to include a list of references you can use the special section title
+`# References`^[it must be exactly this!], and provide a bibtex file in the vignette header.
```
[the remark this answers is not in the document]
Comment (grimbough) [reply vignettes/AuthoringRmdVignettes.Rmd:378]: I think numbering of the references section is up to the user.  I've added a sentence to the vignette explaining this and indicating that you can use `{-}` to suppress the numbering if that's desired.

File: inst/rmarkdown/templates/pdf_document/skeleton/skeleton.Rmd [hunk 4]
```diff
@@ -86,6 +87,8 @@ Apart from referencing figures (Section \@ref(figures)), tables (Section \@ref(t
 
 Footnotes are displayed as side notes on the right margin^[this is a side note entered as a footnote], which has the advantage that they appear close to the place where they are defined.
 
+# References {.unnumbered}
+
 
 # Session info {.unnumbered}
```
[the remark this answers is not in the document]
Comment (grimbough) [reply inst/rmarkdown/templates/pdf_document/skeleton/skeleton.Rmd]: I'm a bit conflicted here.  Assuming everything is left intact then I agree, it's nicer to have the session info as "Appendix A" and it should be numbered.  However, if a vignette doesn't have any references, and someone just deletes the `# References` then "Session Info" will then appear as a regular section and be numbered, which is inconsistent with the old style.  I don't really know what's best - maybe it doesn't really matter as a user can always choose whether the session info section is numbered or not.

For now I've removed the `{.unnumbered}` so the skeleton is consistent with styling of the package vignette itself.

File: inst/rmarkdown/templates/html_document/skeleton/skeleton.Rmd (hunk 2 above)
[the remark this answers is not in the document]
Comment (aoles) [reply inst/rmarkdown/templates/html_document/skeleton/skeleton.Rmd]: I was able to follow up a bit more on this and indeed `link-citations` is a [parameter to pandoc](https://pandoc.org/MANUAL.html#other-relevant-metadata-fields) which apparently is not being handled by `rmarkdown`. Nevertheless, it should be still possible to override the implicit default `link-citation: false` in the `pre_processor` function defined in [rmd_common.R](https://github.com/Bioconductor/BiocStyle/blob/devel/R/rmd_common.R), what do you think?

[the remark this answers is not in the document]
Comment (grimbough) [reply inst/rmarkdown/templates/html_document/skeleton/skeleton.Rmd]: Nice spot. I never spend enough time with the BiocStyle code base to really remember where everything is happening.  I've tried to apply that approach in https://github.com/Bioconductor/BiocStyle/pull/95/commits/ffca11cc1d60f60a4ba057d4c1a2b1a811a78e97

I think that should respect the setting if a user has supplied `link-citations` in the header themselves, and apply `true` if they haven't.  

I've also removed the default entries from the skeleton files.

[the remark this answers is not in the document]
Comment (aoles) [reply inst/rmarkdown/templates/html_document/skeleton/skeleton.Rmd]: Awesome, thanks!

**Review by aoles — APPROVED** [review]

**Comment by aoles** [comment]
Thanks @grimbough for all your nice work! 👍 I appreciate you efforts in maintaining BiocStyle ❤️

**Pull request merged**